src/ape_cache/query.py

```diff
@@ -0,0 +1,169 @@
+                        """
+                        SELECT COUNT(*)
+                        FROM blocks
+                        WHERE blocks.number >= :start_block
+                        AND blocks.number <= :stop_block
+                        AND blocks.number mod :step = 0
+                        """
+                    ),
+                    start_block=query.start_block,
+                    stop_block=query.stop_block,
+                    step=query.step,
+                )
+                number_of_rows = q.rowcount
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
+            number_of_rows = 0
+
+        # NOTE: Assume 200 msec to get data from database
+        time_to_get_cached_records = 200 if number_of_rows > 0 else 0
+        # NOTE: Very loose estimate of 0.75ms per block
+        time_to_get_uncached_records = int(
+            ((query.stop_block - query.start_block) / query.step - number_of_rows) * 0.75
+        )
+        return time_to_get_cached_records + time_to_get_uncached_records
+
+    @singledispatchmethod
+    def perform_query(self, query: QueryType) -> pd.DataFrame:  # type: ignore
+        raise QueryEngineError(f"Cannot handle '{type(query)}'.")
+
+    @perform_query.register
+    def perform_block_query(self, query: BlockQuery) -> pd.DataFrame:
+        try:
+            with self.engine.connect() as conn:
+                q = conn.execute(
+                    text(
+                        """
+                        SELECT :columns
+                        FROM blocks
+                        WHERE blocks.number >= :start_block
+                        AND blocks.number <= :stop_block
+                        AND blocks.number mod :step = 0
+                        """
+                    ),
+                    columns=",".join(query.columns),
+                    start_block=query.start_block,
+                    stop_block=query.stop_block,
+                    step=query.step,
+                )
+                cached_records = pd.DataFrame(columns=query.columns, data=q.fetchall())
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
+            cached_records = pd.DataFrame(columns=query.columns)
+
+        if len(cached_records) == math.ceil((query.stop_block - query.start_block) / query.step):
+            return cached_records
+
+        blocks_iter = map(
+            self.provider.get_block,
+            # NOTE: the range stop block is a non-inclusive stop.
+            #       Where as the query method is an inclusive stop.
+            range(query.start_block + len(cached_records), query.stop_block + 1, query.step),
+        )
+        block_dicts_iter = map(partial(get_columns_from_item, query), blocks_iter)
```

Comment:
This is where the filter is being applied, so it would be like this to back-populate the database:
```suggestion
        # Backfill cache with updates
        blocks_iter, blocks_iter_copy = itertools.tee(blocks_iter)
        query_kwargs = query.dict()
        query_kwargs["columns"] = query.all_fields()
        uncached_unfiltered_records = pd.DataFrame(columns=query.all_fields(), data=block_iter_copy)
        self.update_cache(BlockQuery(**query_kwargs), uncached_unfiltered_records)
        
        block_dicts_iter = map(partial(get_columns_from_item, query), blocks_iter)
```

Reply:
Could probably also refactor this a ton w/ constructing the dataframe and then filtering out the columns afterwards before concating them togther for return

Reply:
I love this solution